### PR TITLE
Implement Transport-Level Fragmentation And Reassembly

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Adds transport-level fragmentation and reassembly to support large payloads without burdening the application layer.
- Fragments are encoded with a FRAG marker, header, and payload, enabling per-fragment framing over the transport.
- Fragmentation defaults are derived from buffer capacity but can be customized or disabled by the application.
- Exposes fragmentation configuration and utilities via the public Fragmentation API.

## Changes
- New fragmentation configuration and encoding/decoding:
  - src/fragment/config.rs: FragmentationConfig with for_frame_budget and encoded_fragment_ceiling helpers.
  - src/fragment/payload.rs: Fragment encoding/decoding helpers (encode_fragment_payload, decode_fragment_payload), plus fragment_overhead and FRAGMENT_MAGIC. Includes unit tests for round-trip and non-fragment payloads.
  - src/fragment/mod.rs: Exposes FragmentationConfig and payload utilities publicly.
- WireframeApp integration:
  - src/app/builder.rs: Added default_fragmentation(capacity) to derive a FragmentationConfig from buffer_capacity; WireframeApp now carries an optional fragmentation config.
  - WireframeApp::fragmentation(...) API added to enable/disable fragmentation; default derives from buffer_capacity.
- Connection and framing behavior:
  - src/app/connection.rs: Introduces FragmentationState (fragmenter + reassembler) and FragmentProcessError handling. Fragmentation is applied to outbound frames when enabled and inbound frames are reassembled into complete envelopes when possible.
  - src/connection.rs: Public API branch to enable fragmentation on outbound frames is added in ConnectionActor (enable_fragmentation).
  - Outbound fragmentation uses Fragmenter to split large payloads into fragment frames and encodes them with the fragment payload encoding utilities.
  - Inbound frames are decoded and reassembled; timeouts and errors propagate as deserialization or fragmentation errors with metrics.
- Tests
  - tests/connection_fragmentation.rs: Verifies outbound fragmentation behavior and reassembly of frames.
  - tests/fragment_transport.rs: End-to-end tests for fragmented request/response, out-of-order fragments, duplicates, and eviction of stale fragments.
- Public API surface
  - src/lib.rs: Exposes FRAGMENT_MAGIC, FragmentationConfig, fragmentation helpers, and encoding/decoding utilities via fragment module.
- Documentation updates (docs/*):
  - docs/generic-message-fragmentation-and-re-assembly-design.md, docs/hardening-wireframe-a-guide-to-production-resilience.md, docs/roadmap.md, and docs/users-guide.md updated to reflect transport-level fragmentation behavior, defaults, and configuration.

## Why
- Enables safe handling of large messages by fragmenting at the transport level, avoiding large allocations or timeouts at the application layer.
- Provides configurable controls for maximum message sizes, fragment sizes, and reassembly timeouts to suit different network environments.
- Keeps fragmentation transparent to application handlers when enabled, preserving existing envelope/message interfaces.

## How to test
- [x] Run cargo test to exercise unit tests for fragmentation payload encoding/decoding.
- [x] Run integration tests in tests/fragment_transport.rs to verify fragmented request/response round-trips.
- [x] Run tests in tests/connection_fragmentation.rs to validate outbound fragmentation behavior and reassembly, including eviction of stale assemblies.
- [x] Manual scenario: configure WireframeApp with fragmentation(Some(cfg)) using a frame budget derived from buffer_capacity and verify fragmented envelopes are produced and reassembled correctly.

## API usage notes
- Enable fragmentation: WireframeApp::new()?.fragmentation(Some(cfg)).route(...)
- Disable fragmentation: WireframeApp::new()?.fragmentation(None).route(...)
- FragmentationConfig is exposed via fragment::FragmentationConfig and can be derived from for_frame_budget(frame_budget, max_message_size, reassembly_timeout).
- Fragmentation utilities available via fragment::encode_fragment_payload, fragment::decode_fragment_payload, and FRAGMENT_MAGIC.

## Migration & compatibility
- Default fragmentation is auto-derived from buffer_capacity unless explicitly overridden with fragmentation(None).
- Existing application handlers receive complete envelopes when fragmentation is enabled; if fragmentation is disabled, behavior remains unchanged.

## Performance considerations
- Fragmentation adds minor overhead for encoding/decoding fragment headers but significantly improves resilience and throughput for large messages in constrained transports.
- Reassembly timeouts and payload caps help bound memory usage and prevent resource leaks from partial fragments.

🌿 Generated by [Terry](https://www.terragonlabs.com)

---

ℹ️ Tag @terragon-labs to ask questions and address PR feedback

📎 **Task**: https://www.terragonlabs.com/task/aff0ff40-67c7-46c3-a989-0fc7f61e7719

## Summary by Sourcery

Add transport-level fragmentation and reassembly so large messages are split into framed fragments on write and transparently reassembled on read, with sensible defaults and configuration hooks.

New Features:
- Introduce FragmentationConfig and fragment payload encoding/decoding helpers to describe and encode transport-level message fragments.
- Enable optional outbound fragmentation and inbound reassembly in WireframeApp and ConnectionActor so handlers continue to work with complete envelopes.
- Expose fragment-related types and helpers (including FRAGMENT_MAGIC) through the public crate API for use by applications and clients.

Enhancements:
- Derive default fragmentation settings from buffer capacity in the WireframeApp builder, with an option to override or disable fragmentation explicitly.
- Refactor connection handling to separate envelope decoding, fragment reassembly, and response forwarding, improving clarity and enabling fragmentation hooks.

Documentation:
- Document the fragment wire encoding, default guards, configuration options, and updated behaviour in the user guide, design doc, hardening guide, and roadmap.

Tests:
- Add integration tests covering fragmented request/response round trips, out-of-order and duplicate fragment handling, and eviction of expired fragments.
- Add a ConnectionActor test to verify outbound frames are fragmented and can be reassembled back into the original payload.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai pause

File: tests/fragment_transport.rs [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+    WireframeApp::new()
+        .expect("build app")
+        .buffer_capacity(capacity)
+        .fragmentation(Some(config))
+        .route(ROUTE_ID, handler)
+        .expect("register route")
+}
+
+#[tokio::test]
+async fn fragmented_request_and_response_round_trip() {
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/fragment_transport.rs]: Has this now been resolved?

File: tests/connection_fragmentation.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+use tokio_util::sync::CancellationToken;
+use wireframe::{
+    ConnectionActor,
+    app::{Envelope, Packet},
+    fragment::{FragmentationConfig, Reassembler, decode_fragment_payload},
+    push::PushQueues,
+};
+
+#[tokio::test]
+async fn connection_actor_fragments_outbound_frames() {
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/connection_fragmentation.rs]: Has this now been resolved?

File: src/fragment/config.rs [hunk 3]
```diff
@@ -0,0 +1,58 @@
+//! Configuration used by transport-level fragmentation and re-assembly.
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/fragment/config.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

File: docs/users-guide.md [hunk 4]
```diff
@@ -308,92 +308,40 @@ async fn main() -> Result<(), SendError> {
...
+not fit into a single frame is split into fragments carrying a `FragmentHeader`
+(`message_id`, `fragment_index`, `is_last_fragment`) wrapped with the `FRAG`
+marker. The connection re-assembles fragments before invoking handlers, so
+handlers continue to work with complete `Envelope` values.[^6]
 
-```rust
-use std::sync::Arc;
+Fragmented messages enforce two guards: `max_message_size` caps the total
+reassembled payload and `reassembly_timeout` evicts stale partial messages.
+Customise or disable fragmentation via the builder:
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/users-guide.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

File: src/app/connection.rs [hunk 5]
```diff
@@ -9,14 +10,25 @@
...
~};
 use tokio_util::codec::{Encoder, Framed, LengthDelimitedCodec};
 
 use super::{
-    builder::WireframeApp,
+    builder::{default_fragmentation, WireframeApp},
     envelope::{Envelope, Packet, PacketParts},
     error::SendError,
+    fragment_utils::fragment_packet,
 };
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/app/connection.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

File: src/app/fragmentation_state.rs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+pub(crate) struct FragmentationState {
+    fragmenter: Fragmenter,
+    reassembler: Reassembler,
+}
+
+/// Decode or reassembly failures encountered while handling fragments.
+pub(crate) enum FragmentProcessError {
+    Decode(DecodeError),
+    Reassembly(ReassemblyError),
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/app/fragmentation_state.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.
